### ISSUE TITLE
S3928 False positive with anonymous method

### ISSUE THREAD

**Issue by lee-11** [SonarSource/sonar-dotnet#3667]
### Description

I have a slew of csharpsquid:S3928 that are false positives when the exception is thrown from within an anonymous method declared within the method body and used as an argument to a method call in the outer method (code below).  This is a common pattern in C# (Linq).

### Repro steps

Please provide a code snippet that compiles which reproduces the problem. Alternatively, a link to a SonarCloud project where the issue appears, or a git repository, or attach a zip with a reproducer project.

### Expected behavior

The following code produces the false positive:

```csharp
void MethodWithInput(string input) {
    InnerMethodWithAction(() => throw new ArgumentException("Input is required", "input"));
}

void InnerMethodWithAction(Action action) {
    action();
}
```

In this situation the exception will be correctly throw from the perspective of the caller to `MethodWithInput` and the issue is not correct.  

### Actual behavior

The code throwing the exception is marked with issue S3938 and labeled "The parameter name 'input' is not declared in the argument list." inaccurately. 

### Known workarounds

Disabling the rule is the only reliable workaround.  In some cases the exception can be thrown from `MethodWithInput` when the input is known to be bad at that point. However, in many cases the disposition of the input isn't known until the inner method is invoked (i.e. the callback is only conditionally used).

### Related information

* dotnet core 3.1
* SonarScanner for MSBuild 4.10
* Amazon Linux 2

**Comment by pavel-mikula-sonarsource** [comment]
Hi @lee-11,

This is actually a True Positive. The reasoning is not so obvious on simple cases. You can check more complex example in [this community thread an answer](https://community.sonarsource.com/t/false-positive-on-c-parameter-names-used-into-argumentexception/17415/21)

**Comment by lee-11** [comment]
That discussion doesn't really address the use of anonymous functions. Given the use case here I think you may want to reconsider whether you believe this is the correct behavior. The ArgumentException is difficult enough to use with its two (often reversed) string parameters, identifying code like this as problematic just makes those cases worse since the rule must be disabled.￼￼

It seems possible to identify this code as less likely problematic, at least, since the method declaration is within a method body with such a named parameter.

**Comment by pavel-mikula-sonarsource** [comment]
It really is the same example. Let me rewrite it to make it more obvious for your sample:
```C#
public void MethodWithInput(string input)
{
    InnerMethodWithAction(() => throw new ArgumentException("Input is required", "input"));
}

void InnerMethodWithAction(Action action)
{
    LastOne(action, "Something");
}

void LastOne(Action action, string input)
{
    if(input != null)
    {
        action();
    }
}
```
This piece of code will raise `ArgumentException` with `Input is required` message right at the place, where we're 100% sure that `input` argument is not null at all.
